File: frontend/src/types/continuity.ts
import type { PaginatedData } from "./api";

export interface CharacterState {
  id: string;
  project_id: string;
  character_name: string;
  episode_start: number;
  episode_end: number | null;
  outfit_description: string;
  hairstyle: string;
  accessories: Record<string, any>;
  makeup: string;
  age_appearance: string;
  lora_path: string | null;
  embedding_path: string | null;
  reference_image_path: string | null;
  signature_items: Record<string, any>;
  notes: string;
  created_at: string;
  updated_at: string;
}

export interface CharacterStateCreate {
  project_id: string;
  character_name: string;
  episode_start: number;
  episode_end?: number | null;
  outfit_description?: string;
  hairstyle?: string;
  accessories?: Record<string, any>;
  makeup?: string;
  age_appearance?: string;
  lora_path?: string | null;
  embedding_path?: string | null;
  reference_image_path?: string | null;
  signature_items?: Record<string, any>;
  notes?: string;
}

export interface CharacterStateUpdate {
  episode_end?: number | null;
  outfit_description?: string;
  hairstyle?: string;
  accessories?: Record<string, any>;
  makeup?: string;
  age_appearance?: string;
  lora_path?: string | null;
  embedding_path?: string | null;
  reference_image_path?: string | null;
  signature_items?: Record<string, any>;
  notes?: string;
}

export interface SceneAsset {
  id: string;
  project_id: string;
  scene_name: string;
  scene_type: "interior" | "exterior" | "mixed";
  description: string;
  layout_description: string;
  lora_path: string | null;
  controlnet_depth_path: string | null;
  controlnet_edge_path: string | null;
  variants: Record<string, any>;
  reference_image_path: string | null;
  created_at: string;
  updated_at: string;
}

export interface SceneAssetCreate {
  project_id: string;
  scene_name: string;
  scene_type?: "interior" | "exterior" | "mixed";
  description?: string;
  layout_description?: string;
  lora_path?: string | null;
  controlnet_depth_path?: string | null;
  controlnet_edge_path?: string | null;
  variants?: Record<string, any>;
  reference_image_path?: string | null;
}

export interface SceneAssetUpdate {
  scene_name?: string;
  scene_type?: "interior" | "exterior" | "mixed";
  description?: string;
  layout_description?: string;
  lora_path?: string | null;
  controlnet_depth_path?: string | null;
  controlnet_edge_path?: string | null;
  variants?: Record<string, any>;
  reference_image_path?: string | null;
}

export interface PacingTemplate {
  id: string;
  name: string;
  description: string;
  genre: string;
  structure: Record<string, any>;
  hook_3sec_config: Record<string, any>;
  cliffhanger_config: Record<string, any>;
  target_duration_sec: number;
  usage_count: number;
  avg_completion_rate: number;
  created_at: string;
  updated_at: string;
}

export interface PacingTemplateCreate {
  name: string;
  description?: string;
  genre?: string;
  structure?: Record<string, any>;
  hook_3sec_config?: Record<string, any>;
  cliffhanger_config?: Record<string, any>;
  target_duration_sec?: number;
}

export interface PacingValidationRequest {
  scene_content: Record<string, any>;
  template_id?: string;
  hook_config?: Record<string, any>;
}

export interface PacingValidationResult {
  valid: boolean;
  issues: Array<{
    type: string;
    message: string;
    severity: "error" | "warning" | "info";
  }>;
}

export interface ComplianceReport {
  id: string;
  project_id: string;
  check_type: "face_recognition" | "music_copyright" | "content_moderation";
  episode_number: number | null;
  stage_type: string | null;
  status: "pending" | "completed" | "failed";
  violations: number;
  violations_detail: Record<string, any>;
  checked_at: string;
}

export interface ComplianceCheckRequest {
  project_id: string;
  check_type: "face_recognition" | "music_copyright" | "content_moderation";
  episode_number?: number;
  stage_type?: string;
}

export interface AsyncTaskResponse {
  task_id: string;
  status: string;
  message: string;
}

export interface ConsistencyCheck {
  id: string;
  project_id: string;
  check_type: string;
  episode_start: number;
  episode_end: number;
  status: "pending" | "completed" | "failed";
  issues_found: number;
  issues_detail: Record<string, any>;
  checked_at: string;
}

export interface ConsistencyCheckRequest {
  project_id: string;
  check_type: string;
  episode_start: number;
  episode_end: number;
}

File: backend/app/schemas/continuity.py
"""
连续性管理相关 Schema。
"""

from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, Field


class CharacterStateCreate(BaseModel):
    project_id: UUID
    character_name: str = Field(min_length=1, max_length=100)
    episode_start: int = Field(ge=1)
    episode_end: int | None = Field(default=None, ge=1)
    outfit_description: str = Field(default="", max_length=2000)
    hairstyle: str = Field(default="", max_length=200)
    accessories: dict[str, Any] = Field(default_factory=dict)
    makeup: str = Field(default="", max_length=200)
    age_appearance: str = Field(default="", max_length=100)
    lora_path: str | None = Field(default=None, max_length=500)
    embedding_path: str | None = Field(default=None, max_length=500)
    reference_image_path: str | None = Field(default=None, max_length=500)
    signature_items: dict[str, Any] = Field(default_factory=dict)
    notes: str = Field(default="", max_length=2000)


class CharacterStateResponse(BaseModel):
    id: UUID
    project_id: UUID
    character_name: str
    episode_start: int
    episode_end: int | None
    outfit_description: str
    hairstyle: str
    accessories: dict[str, Any]
    makeup: str
    age_appearance: str
    lora_path: str | None
    embedding_path: str | None
    reference_image_path: str | None
    signature_items: dict[str, Any]
    notes: str
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class SceneAssetCreate(BaseModel):
    project_id: UUID
    scene_name: str = Field(min_length=1, max_length=200)
    scene_type: str = Field(default="interior", max_length=50)
    description: str = Field(default="", max_length=2000)
    layout_description: str = Field(default="", max_length=2000)
    lora_path: str | None = Field(default=None, max_length=500)
    controlnet_depth_path: str | None = Field(default=None, max_length=500)
    controlnet_edge_path: str | None = Field(default=None, max_length=500)
    variants: dict[str, Any] = Field(default_factory=dict)
    reference_image_path: str | None = Field(default=None, max_length=500)


class SceneAssetResponse(BaseModel):
    id: UUID
    project_id: UUID
    scene_name: str
    scene_type: str
    description: str
    layout_description: str
    lora_path: str | None
    controlnet_depth_path: str | None
    controlnet_edge_path: str | None
    variants: dict[str, Any]
    reference_image_path: str | None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class ConsistencyCheckResponse(BaseModel):
    id: UUID
    project_id: UUID
    check_type: str
    episode_start: int
    episode_end: int
    status: str
    issues_found: int
    issues_detail: dict[str, Any]
    checked_at: datetime

    model_config = {"from_attributes": True}


class PacingTemplateCreate(BaseModel):
    name: str = Field(min_length=1, max_length=200)
    description: str = Field(default="", max_length=2000)
    genre: str = Field(default="", max_length=100)
    structure: dict[str, Any] = Field(default_factory=dict)
    hook_3sec_config: dict[str, Any] = Field(default_factory=dict)
    cliffhanger_config: dict[str, Any] = Field(default_factory=dict)
    target_duration_sec: int = Field(default=60, ge=10, le=600)


class PacingTemplateResponse(BaseModel):
    id: UUID
    name: str
    description: str
    genre: str
    structure: dict[str, Any]
    hook_3sec_config: dict[str, Any]
    cliffhanger_config: dict[str, Any]
    target_duration_sec: int
    usage_count: int
    avg_completion_rate: float
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}